File: generate.py
"""github-pet generator: builds dist/*.svg from live GitHub data.

Merged architecture:
- state machine (pet/state.py) fed by public events (REST) + GraphQL activity
- real contribution calendar via GraphQL (pet/graph_api.py)
- banner scene: the cat lives a little daily routine (pet/render.py)
- hero visual: mochi kitty hopping along the contribution graph (pet/graph_render.py)
- language chart from real repo language bytes (pet/charts.py)
"""
from __future__ import annotations

import os
import pathlib

from pet.github_api import fetch_events, fetch_repos, fetch_languages
from pet.state import decide
from pet import charts, graph_api
from pet.graph_render import build_graph_svg
from pet.render import build_svg, PALETTES

USER = os.environ.get("PET_USER", "prsdx")


def main() -> None:
    token = os.environ.get("GITHUB_TOKEN") or os.environ.get("PET_GITHUB_TOKEN")
    events = fetch_events(USER)
    repos = fetch_repos(USER)
    langs = fetch_languages(USER, repos)
    calendar = graph_api.fetch_calendar(token, USER)
    activity = graph_api.fetch_activity(token, USER) or {}
    status = decide(events, activity=activity)

    dist = pathlib.Path("dist")
    dist.mkdir(exist_ok=True)
    outputs = {
        "pet.svg": build_svg(status.state, status.caption, "dark"),
        "pet-light.svg": build_svg(status.state, status.caption, "light"),
        "graph.svg": build_graph_svg(status.state, status.caption, calendar, "dark"),
        "graph-light.svg": build_graph_svg(status.state, status.caption, calendar, "light"),
        "langs.svg": charts.langs_chart(langs, len(repos), PALETTES["dark"]),
        "langs-light.svg": charts.langs_chart(langs, len(repos), PALETTES["light"]),
    }
    for name, svg in outputs.items():
        (dist / name).write_text(svg, encoding="utf-8")
        print(f"wrote dist/{name} ({len(svg)} bytes)")
    cal = f"{calendar['total']} contributions" if calendar else "unavailable"
    print(f"state={status.state} | caption={status.caption!r} | api_ok={status.api_ok}")
    print(f"calendar={cal} | last_push={activity.get('last_push')} | events={len(events)} repos={len(repos)} langs={len(langs)}")


if __name__ == "__main__":
    main()

File: generate.ts
// github-pet generator: builds dist/*.svg from live GitHub data.
// Zero-dependency TypeScript - runs on Bun (CI) or Node 24+ (type stripping).

import { mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { fetchEvents, fetchRepos, fetchLanguages, fetchCiStatus, fetchProfile } from "./src/githubApi.ts";
import { fetchCalendar, fetchActivity } from "./src/graphApi.ts";
import { decide, greetingFor } from "./src/state.ts";
import { buildSvg, PALETTES } from "./src/render.ts";
import { buildGraphSvg } from "./src/graphRender.ts";
import { buildIsoSvg } from "./src/isoRender.ts";
import { langsChart } from "./src/charts.ts";

const USER = process.env.PET_USER || "prsdx";
const OUT_DIR = process.env.PET_OUTPUT_DIR || "dist";
const CONTACT = process.env.PET_CONTACT || "";
const ATTRIBUTION = !["0", "false", "no"].includes((process.env.PET_ATTRIBUTION ?? "").toLowerCase());

async function main(): Promise<void> {
    const token = process.env.GITHUB_TOKEN || process.env.PET_GITHUB_TOKEN;
    const [events, repos] = await Promise.all([fetchEvents(USER), fetchRepos(USER)]);
    const langs = await fetchLanguages(USER, repos);
    const [calendar, activity] = await Promise.all([fetchCalendar(token, USER), fetchActivity(token, USER)]);
    const ci = await fetchCiStatus(USER);
    const status = decide(events, {
        lastPush: activity?.lastPush ?? null,
        merged24h: activity?.merged24h ?? 0,
    }, ci);

    mkdirSync(OUT_DIR, { recursive: true });
    const name = process.env.PET_NAME || (await fetchProfile(USER))?.name?.split(" ")[0]?.toLowerCase() || "";
    const greeting = greetingFor(new Date(), name);
    const outputs: Record<string, string> = {
        "pet.svg": buildSvg(status.state, status.caption, "dark", greeting, CONTACT, ATTRIBUTION),
        "pet-light.svg": buildSvg(status.state, status.caption, "light", greeting, CONTACT, ATTRIBUTION),
        "graph.svg": buildGraphSvg(status.state, status.caption, calendar, "dark", ATTRIBUTION),
        "graph-light.svg": buildGraphSvg(status.state, status.caption, calendar, "light", ATTRIBUTION),
        "isocat.svg": buildIsoSvg(status.state, status.caption, calendar, "dark", ATTRIBUTION),
        "isocat-light.svg": buildIsoSvg(status.state, status.caption, calendar, "light", ATTRIBUTION),
        "langs.svg": langsChart(langs, repos.length, PALETTES.dark),
        "langs-light.svg": langsChart(langs, repos.length, PALETTES.light),
    };
    for (const [name, svg] of Object.entries(outputs)) {
        writeFileSync(join(OUT_DIR, name), svg, "utf-8");
        console.log(`wrote ${OUT_DIR}/${name} (${svg.length} bytes)`);
    }
    console.log(`state=${status.state} | caption='${status.caption}' | apiOk=${status.apiOk} | ci_failed=${ci.failed}`);
    console.log(`calendar=${calendar ? calendar.total + " contributions" : "unavailable"} | lastPush=${activity?.lastPush ?? "?"} | events=${events.length} repos=${repos.length} langs=${Object.keys(langs).length}`);
}

main();